include/envoy/upstream/upstream.h

```diff
@@ -168,6 +171,70 @@ class HostSet {
    * @return same as hostsPerLocality but only contains healthy hosts.
    */
   virtual const std::vector<std::vector<HostSharedPtr>>& healthyHostsPerLocality() const PURE;
+
+  /**
+   * Updates the hosts in a given host set.
+   *
+   * @param hosts supplies the (usually new) list of hosts in the host set.
+   * @param healthy hosts supplies the subset of hosts which are healthy.
+   * @param hosts_per_locality supplies the hosts subdivided by locality.
+   * @param hosts_per_locality supplies the healthy hosts subdivided by locality.
+   * @param hosts_added supplies the hosts added since the last update.
+   * @param hosts_removed supplies the hosts removed since the last update.
+   */
+  virtual void updateHosts(
+      std::shared_ptr<const std::vector<HostSharedPtr>> hosts,
+      std::shared_ptr<const std::vector<HostSharedPtr>> healthy_hosts,
+      std::shared_ptr<const std::vector<std::vector<HostSharedPtr>>> hosts_per_locality,
+      std::shared_ptr<const std::vector<std::vector<HostSharedPtr>>> healthy_hosts_per_locality,
+      const std::vector<HostSharedPtr>& hosts_added,
+      const std::vector<HostSharedPtr>& hosts_removed) PURE;
+
+  /**
+   * @return the priority of this host set.
+   */
+  virtual uint32_t priority() const PURE;
+};
+
+typedef std::unique_ptr<HostSet> HostSetPtr;
+
+/**
+ * This class contains all of the HostSets for a given cluster grouped by priority, for
+ * ease of load balancing.
+ */
+class PrioritySet {
+public:
+  typedef std::function<void(uint32_t priority, const std::vector<HostSharedPtr>& hosts_added,
+                             const std::vector<HostSharedPtr>& hosts_removed)>
+      MemberUpdateCb;
+
+  virtual ~PrioritySet() {}
+
+  /**
+   * Install a callback that will be invoked when any of the HostSets in the PrioritySet changes.
+   * This includes when a new HostSet is created.
+   *
+   * @param callback supplies the callback to invoke.
+   */
+  virtual Common::CallbackHandle* addMemberUpdateCb(MemberUpdateCb callback) const PURE;
+
+  /**
+   * Returns the host sets for this priority set, ordered by priority.
+   * The first element in the vector is the host set for priority 0, and so on.
+   *
+   * @return the host sets for this priority set.
+   */
+  virtual std::vector<HostSetPtr>& hostSetsPerPriority() PURE;
+
+  /**
+   * @return the const host sets for this priority set, ordered by priority.
+   */
+  virtual const std::vector<HostSetPtr>& hostSetsPerPriority() const PURE;
+
+  /**
+   * @return the host sets for this priority.  If no such host set exists, it will be created.
+   */
+  virtual HostSet& getHostSet(uint32_t priority) PURE;
```

Comment:
Do we ever remove hosts sets at a given priority?

Reply:
Nope.  I thought that was more of an implementation detail though so that's commented upstream_impl.h